File: java/src/main/java/com/company/automation/support/PlaywrightFactory.java
package com.company.automation.support;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserContext;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.Playwright;
import com.microsoft.playwright.options.ViewportSize;
import java.nio.file.Path;
import java.util.Arrays;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Owns the Playwright lifecycle.
 *
 * <p>Two constraints drive the design:
 *
 * <ul>
 *   <li>Playwright objects are <b>not thread safe</b>. With JUnit parallel execution enabled, each
 *       thread needs its own {@link Playwright} and {@link Browser}, so both are {@link
 *       ThreadLocal}.
 *   <li>{@code Playwright.create()} launches a Node process and costs seconds. Creating one per
 *       test would dominate the suite runtime, so it is created once per thread and reused.
 * </ul>
 *
 * <p>A {@link BrowserContext} is cheap and gives complete isolation: separate cookies, storage, and
 * cache. That is why each test gets a fresh one rather than sharing a page. Sharing a page across
 * tests is the cause of the classic "fails only when it runs second" flake.
 */
public final class PlaywrightFactory {

  private static final Logger log = LoggerFactory.getLogger(PlaywrightFactory.class);

  private static final ThreadLocal<Playwright> PLAYWRIGHT = new ThreadLocal<>();
  private static final ThreadLocal<Browser> BROWSER = new ThreadLocal<>();

  private PlaywrightFactory() {}

  /** The browser for the current thread, launched on first use. */
  public static Browser browser() {
    Browser existing = BROWSER.get();
    if (existing != null) {
      return existing;
    }

    Config config = ConfigLoader.config();
    Playwright playwright = Playwright.create();

    // Applied once per Playwright instance. The attribute is per-application, so it comes from
    // config rather than being hardcoded: this repo's demo target uses data-test, not data-testid.
    playwright.selectors().setTestIdAttribute(config.execution().testIdAttribute());

    BrowserType browserType = browserTypeFor(playwright, config.execution().browser());
    BrowserType.LaunchOptions options = new BrowserType.LaunchOptions();
    options.setHeadless(config.execution().headless());

    // Disabling animations removes a whole class of "fails as the modal appears" flake.
    options.setArgs(Arrays.asList("--force-prefers-reduced-motion"));

    Browser launched = browserType.launch(options);

    PLAYWRIGHT.set(playwright);
    BROWSER.set(launched);

    log.debug(
        "Launched {} (headless={}) on thread {}",
        config.execution().browser(),
        config.execution().headless(),
        Thread.currentThread().getName());

    return launched;
  }

  /**
   * A fresh, isolated context for one test, with tracing already started.
   *
   * <p>Tracing is always started and only ever <i>saved</i> on failure. Starting it conditionally
   * would mean the first failure is the one run with no trace, which is precisely the run you need.
   */
  public static BrowserContext newContext() {
    Config config = ConfigLoader.config();

    Browser.NewContextOptions options = new Browser.NewContextOptions();
    options.setViewportSize(
        new ViewportSize(config.ui().viewportWidth(), config.ui().viewportHeight()));
    options.setLocale(config.ui().locale());
    options.setTimezoneId(config.ui().timezoneId());
    options.setBaseURL(config.ui().baseUrl());

    BrowserContext context = browser().newContext(options);
    context.setDefaultTimeout(config.timeouts().elementMs());
    context.setDefaultNavigationTimeout(config.timeouts().navigationMs());

    context
        .tracing()
        .start(
            new com.microsoft.playwright.Tracing.StartOptions()
                .setScreenshots(true)
                .setSnapshots(true)
                .setSources(true));

    return context;
  }

  /** Stops tracing, writing the trace file only when {@code tracePath} is non-null. */
  public static void stopTracing(BrowserContext context, Path tracePath) {
    com.microsoft.playwright.Tracing.StopOptions options =
        new com.microsoft.playwright.Tracing.StopOptions();
    if (tracePath != null) {
      options.setPath(tracePath);
    }
    context.tracing().stop(options);
  }

  /** Opens a page in the given context with the configured timeouts already applied. */
  public static Page newPage(BrowserContext context) {
    return context.newPage();
  }

  /**
   * Closes and clears this thread's browser. Removing the {@link ThreadLocal} matters: leaving it
   * set leaks a Node process per thread for the life of the JVM.
   */
  public static void closeForCurrentThread() {
    Browser browser = BROWSER.get();
    if (browser != null) {
      browser.close();
    }
    BROWSER.remove();

    Playwright playwright = PLAYWRIGHT.get();
    if (playwright != null) {
      playwright.close();
    }
    PLAYWRIGHT.remove();
  }

  private static BrowserType browserTypeFor(Playwright playwright, String name) {
    if ("firefox".equalsIgnoreCase(name)) {
      return playwright.firefox();
    }
    if ("webkit".equalsIgnoreCase(name)) {
      return playwright.webkit();
    }
    if ("chromium".equalsIgnoreCase(name)) {
      return playwright.chromium();
    }
    throw new IllegalArgumentException(
        "Unsupported browser '" + name + "'. Use chromium, firefox or webkit.");
  }
}

File: csharp/src/AutomationFramework.Core/Support/PlaywrightFactory.cs
using Microsoft.Playwright;

namespace AutomationFramework.Core.Support;

/// <summary>
/// Owns the Playwright lifecycle.
/// </summary>
/// <remarks>
/// Two constraints drive the design:
/// <list type="bullet">
///   <item>
///     <c>Playwright.CreateAsync()</c> starts a Node driver process and costs seconds. Doing it per
///     test would dominate the run, so one <see cref="IPlaywright"/> and one <see cref="IBrowser"/>
///     are shared for the whole assembly.
///   </item>
///   <item>
///     An <see cref="IBrowserContext"/> is cheap and gives complete isolation: separate cookies,
///     storage and cache. That is why each test gets a fresh one. Sharing a page across tests is the
///     cause of the classic "fails only when it runs second" flake.
///   </item>
/// </list>
/// <para>
/// Unlike the Java binding, Playwright for .NET is safe to use from multiple threads against one
/// <see cref="IBrowser"/>, so no thread-local is needed here. The isolation boundary is the context,
/// not the thread.
/// </para>
/// </remarks>
public static class PlaywrightFactory
{
    private static IPlaywright? _playwright;
    private static IBrowser? _browser;

    /// <summary>
    /// Creates the shared Playwright instance and browser. Call exactly once, from assembly-level setup.
    /// </summary>
    /// <remarks>
    /// Eager initialization here rather than lazily on first use, and the reason is not style.
    /// <para>
    /// Creating Playwright lazily inside a test's <c>async Task [SetUp]</c> deadlocks. NUnit runs async
    /// setup through an adapter that blocks the worker thread without pumping a message loop, while
    /// <c>Playwright.CreateAsync()</c> has internal continuations that post back to the captured
    /// <see cref="SynchronizationContext"/>. Those continuations then never run, and the suite hangs with no
    /// output at all — not even a test name, which makes it look like an environment failure rather than a
    /// deadlock.
    /// </para>
    /// <para>
    /// <c>Task.Run</c> is what makes this safe: it detaches the work from any ambient synchronization
    /// context, so every continuation lands on the thread pool. Doing it once, before the parallel workers
    /// start, also means the expensive driver start-up is not raced by four threads.
    /// </para>
    /// </remarks>
    public static async Task InitializeAsync()
    {
        if (_browser is not null)
        {
            return;
        }

        AppConfig config = ConfigLoader.Config;
        BrowserInstaller.EnsureInstalled(config.Execution.Browser);

        (IPlaywright playwright, IBrowser browser) = await Task.Run(async () =>
        {
            IPlaywright created = await Playwright.CreateAsync().ConfigureAwait(false);

            // Applied once. The attribute is per-application, so it comes from config rather than being
            // hardcoded: this repo's demo target uses data-test, not data-testid.
            created.Selectors.SetTestIdAttribute(config.Execution.TestIdAttribute);

            IBrowserType browserType = BrowserTypeFor(created, config.Execution.Browser);

            IBrowser launched = await browserType.LaunchAsync(new BrowserTypeLaunchOptions
            {
                Headless = config.Execution.Headless,

                // Disabling animations removes a whole class of "fails as the modal appears" flake.
                Args = ["--force-prefers-reduced-motion"],
            }).ConfigureAwait(false);

            return (created, launched);
        }).ConfigureAwait(false);

        _playwright = playwright;
        _browser = browser;

        TestLog.Info(
            $"Launched {config.Execution.Browser} (headless={config.Execution.Headless}).");
    }

    /// <summary>The shared browser. <see cref="InitializeAsync"/> must have run first.</summary>
    public static IBrowser Browser => _browser
        ?? throw new InvalidOperationException(
            "The browser has not been created. GlobalSetup.BeforeAllTests must call "
            + "PlaywrightFactory.InitializeAsync() before any UI test runs.");

    /// <summary>
    /// A fresh, isolated context for one test, with tracing already started.
    /// </summary>
    /// <remarks>
    /// Tracing is always started and only ever <i>saved</i> on failure. Starting it conditionally would
    /// mean the first failure is the one run with no trace, which is precisely the run you need.
    /// </remarks>
    public static async Task<IBrowserContext> NewContextAsync()
    {
        AppConfig config = ConfigLoader.Config;

        IBrowserContext context = await Browser.NewContextAsync(new BrowserNewContextOptions
        {
            ViewportSize = new ViewportSize
            {
                Width = config.Ui.ViewportWidth,
                Height = config.Ui.ViewportHeight,
            },
            Locale = config.Ui.Locale,
            TimezoneId = config.Ui.TimezoneId,
            BaseURL = config.Ui.BaseUrl,
        }).ConfigureAwait(false);

        context.SetDefaultTimeout(config.Timeouts.ElementMs);
        context.SetDefaultNavigationTimeout(config.Timeouts.NavigationMs);

        await context.Tracing.StartAsync(new TracingStartOptions
        {
            Screenshots = true,
            Snapshots = true,
            Sources = true,
        }).ConfigureAwait(false);

        return context;
    }

    /// <summary>
    /// Stops tracing, writing the trace file only when <paramref name="tracePath"/> is set.
    /// </summary>
    /// <remarks>
    /// Tolerant of an already-closed context. When a test fails because the browser died, stopping the trace
    /// throws as well — and a teardown that throws replaces the real failure in the report with a confusing
    /// secondary one.
    /// </remarks>
    public static async Task StopTracingAsync(IBrowserContext context, string? tracePath)
    {
        try
        {
            await context.Tracing.StopAsync(new TracingStopOptions { Path = tracePath })
                .ConfigureAwait(false);
        }
        catch (PlaywrightException e)
        {
            TestLog.Warn($"Could not stop tracing: {e.Message}");
        }
    }

    /// <summary>Closes the shared browser. Called once from the assembly-level teardown.</summary>
    public static async Task CloseAsync()
    {
        if (_browser is not null)
        {
            await _browser.CloseAsync().ConfigureAwait(false);
            _browser = null;
        }

        _playwright?.Dispose();
        _playwright = null;
    }

    private static IBrowserType BrowserTypeFor(IPlaywright playwright, string name)
    {
        if (string.Equals(name, "firefox", StringComparison.OrdinalIgnoreCase))
        {
            return playwright.Firefox;
        }

        if (string.Equals(name, "webkit", StringComparison.OrdinalIgnoreCase))
        {
            return playwright.Webkit;
        }

        if (string.Equals(name, "chromium", StringComparison.OrdinalIgnoreCase))
        {
            return playwright.Chromium;
        }

        throw new ArgumentException(
            $"Unsupported browser '{name}'. Use chromium, firefox or webkit.", nameof(name));
    }
}
